bson/bsoncodec/empty_interface_codec.go

```diff
@@ -56,24 +52,42 @@ func (eic EmptyInterfaceCodec) EncodeValue(ec EncodeContext, vw bsonrw.ValueWrit
 	return encoder.EncodeValue(ec, vw, val.Elem())
 }
 
+func (eic EmptyInterfaceCodec) getEmptyInterfaceDecodeType(dc DecodeContext, valueType bsontype.Type) (reflect.Type, error) {
+	isDocument := valueType == bsontype.Type(0) || valueType == bsontype.EmbeddedDocument
+	if isDocument && dc.Ancestor != nil {
```

Comment:
This is a logic change. Instead of using ancestor information as a fallback if no type map entry is found, I think we should always use ancestor information if it's present. This is helpful if we're decoding into something like `bson.D`, where the values are of type `interface{}`. Even if the user has registered a type map entry for bsontype.EmbeddedDocument, subdocuments should still decode as `bson.D` because the top-level type was `bson.D`.

Reply:
+1... I argued for this at one point and got shut down. I truly believe this is the right thing to do.